resources/assets/lib/beatmap-discussions/editor-insertion-menu.tsx

```diff
@@ -0,0 +1,295 @@
+
+  // updates cascade from our parent (slate editor), i.e. `componentDidUpdate` gets called on editor changes (typing/selection changes/etc)
+  componentDidUpdate() {
+    this.forceHideMenu();
+  }
+
+  componentWillUnmount() {
+    $(window).off('.editor-insert-menu');
+    if (this.scrollContainer) {
+      $(this.scrollContainer).off('.editor-insert-menu');
+    }
+    if (this.insertRef.current) {
+      $(this.insertRef.current).off('.editor-insert-menu');
+    }
+  }
+
+  containerMouseMove(event: JQuery.MouseMoveEvent) {
+    const block = event.target.closest(`.${editorClass}__block`);
+
+    if (
+      !block ||
+      !this.insertRef.current ||
+      !event.originalEvent ||
+      event.originalEvent.buttons > 0 // don't show while dragging
+    ) {
+      this.hoveredBlock = undefined;
+      return;
+    }
+
+    this.hoveredBlock = block;
+    const blockRect = block.getBoundingClientRect();
+    const cursorPos = {
+      x: event.originalEvent.clientX,
+      y: event.originalEvent.clientY,
+    };
+
+    // If we're past the half-way point of the block's height then put the menu below the block, otherwise put it above
+    if ((cursorPos?.y - blockRect.top) > (blockRect.height / 2)) {
+      this.menuPos = 'below';
+    } else {
+      this.menuPos = 'above';
+    }
+
+    this.updatePosition();
+    this.showMenu();
+    this.startHideTimer();
+  }
+
+  forceHideMenu() {
+    this.mouseOver = false;
+    this.hideMenu();
+  }
+
+  getBlockFromInsertMarker() {
+    const container = this.scrollContainer;
+    const insertMarker = this.insertRef.current;
+    if (!container || !insertMarker) {
+      return;
+    }
+
+    const y = insertMarker.getBoundingClientRect().y + (insertMarker.getBoundingClientRect().height / 2);
+
+    let blockOffset = -1;
+    for (const child of container.children[0].children) {
+      if (y < child.getBoundingClientRect().top) {
+        const prevBlock = container.children[0].children[blockOffset];
+        if (y > prevBlock.getBoundingClientRect().top + (prevBlock.getBoundingClientRect().height / 2)) {
+          blockOffset++;
+        }
+
+        break;
+      }
+      blockOffset++;
+    }
+
+    return container.children[0].children[blockOffset];
+  }
+
+  hideMenu() {
+    if (!this.insertRef.current || this.mouseOver) {
+      return;
+    }
+
+    this.insertRef.current.style.display = 'none';
+  }
+
+  insertBlock = (event: React.MouseEvent<HTMLElement>) => {
+    const e: ReactEditor = this.context;
+    const lastChild = this.getBlockFromInsertMarker()?.lastChild;
+
+    if (!lastChild) {
+      return;
+    }
+
+    let node = ReactEditor.toSlateNode(e, lastChild);
+    let at: Point;
+
+    // TODO: This is a workaround for Slate incorrectly inserting nodes _after_ an empty element instead of _before_.
+    //  Either due to a bug in SlateEditor.end() or with how our 'embed' blocks are implemented... maybe we should
+    //  look at converting the embeds to voids at some point?
```

Comment:
I think another workaround is needed for empty paragraph?

![](https://s.myconan.net/2020-07/2020-07-06_20-20-55.gif)